app/controllers/api/v2/compute_resources_controller.rb

```diff
@@ -148,19 +148,15 @@ def available_security_groups
       api :PUT, "/compute_resources/:id/associate/", N_("Associate VMs to Hosts")
       param :id, :identifier, :required => true
       def associate
+        associator = ComputeResourceHostAssociator.new(@compute_resource)
         @hosts = []
-        if @compute_resource.respond_to?(:associated_host)
-          @compute_resource.vms(:eager_loading => true).each do |vm|
-            if Host.for_vm(@compute_resource, vm).empty?
-              host = @compute_resource.associated_host(vm)
-              if host.present?
-                host.associate!(@compute_resource, vm)
-                @hosts << host
-              end
-            end
-          end
+        if associator.supported_compute_resource?
```

Comment:
Nit: I think, I'd make this method part of the compute resource: `@compute_resource.supports_host_association?` or something.